### PR TITLE
Implements #325 rejectVersionIf { ... }

### PR DESCRIPTION
**Pull request by jmfayard**
Replaces my first experiment, using what I learned from it and from your comments https://github.com/ben-manes/gradle-versions-plugin/pull/337

Also
 
- I updated the README with the new approach of validating what is a stable version
- I define a SAM interface that both a Groovy closure and a Kotlin function can implements properly
- I noticed that in `ComponentSelectionWithCurrent`, the `@delegate` syntactic sugar does not work well in Kotlin so I implemented it manually

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -85,48 +85,112 @@ The strategy can be specified either on the task or as a system property for ad
...
-dependencyUpdates.resolutionStrategy {
-  componentSelection { rules ->
-    rules.all { ComponentSelection selection ->
-      def isNonStable = { String version ->
-        ['alpha', 'beta', 'rc', 'cr', 'm', 'preview', 'b', 'ea'].any { qualifier ->
-          version ==~ /(?i).*[.-]$qualifier[.\d-+]*/
-        }
-      }
+def isNonStable = { String version ->
+  def stableKeyword = ['RELEASE', 'FINAL', 'GA'].any { it -> version.toUpperCase().contains(it) }
```
Comment (ben-manes) [inline README.md]: I don't recall if we filter these out, but `null` version is valid. It most often happens for local file dependencies (via `flatDir`). So this could NPE.

File: README.md [hunk 2]
```diff
@@ -85,48 +85,112 @@ The strategy can be specified either on the task or as a system property for ad
...
+  def regex = /^[0-9,.v-]+$/
```
Comment (ben-manes) [inline README.md]: unused

File: README.md [hunk 3]
```diff
@@ -85,48 +85,112 @@ The strategy can be specified either on the task or as a system property for ad
...
+
+   // Example 2: disallow release candidates as upgradable versions from stable versions
+   rejectVersionIf { selection ->
+       isNonStable(candidate.version) && !isNonStable(currentVersion)
+   }
+
+   // Example 3: using the full syntax
+   resolutionStrategy {
+     componentSelection { rules ->
+       rules.all { ComponentSelection selection ->
```
Comment (ben-manes) [inline README.md]: you can drop the params and use the `delegate` now that it works, making the code look cleaner

File: src/main/groovy/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentSelectionWithCurrent.groovy [hunk 4]
```diff
@@ -1,13 +1,48 @@
...
+    return delegate.getDescriptor(descriptorClass)
+  }
+
+  void reject(String reason) {
+    delegate.reject(reason)
+  }
+
+
+  @Override
+  public String toString() {
```
Comment (ben-manes) [inline src/main/groovy/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentSelectionWithCurrent.groovy]: You can probably use interpolation, e.g.
```groovy
"ComponentSelectionWithCurrent{group=${candidate.group}...}"
```

File: README.md (hunk 1 above)
Comment (jmfayard) [reply README.md]: I would prefer that we filter out null in this case, it would be annoying in Kotlin to have to use a nullable type

File: README.md (hunk 2 above)
Comment (jmfayard) [reply README.md]: fixed

File: README.md (hunk 3 above)
Comment (jmfayard) [reply README.md]: done

File: src/main/groovy/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentSelectionWithCurrent.groovy (hunk 4 above)
Comment (jmfayard) [reply src/main/groovy/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentSelectionWithCurrent.groovy]: done

File: README.md (hunk 1 above)
Comment (ben-manes) [reply README.md]: True, `null` values are annoying. However you are removing a Gradle feature by doing that, which is also undesirable.

Comment (ben-manes) [reply README.md]: Obviously we could show this and assume someone using a local repository would know enough to add the null check themselves. But they might not if blindly copying samples...

**Comment by jmfayard** [comment]
Also: can I rename the samples from `examples` and `examples/kotlin` to `sample-groovy` and `sample-kotlin`? And should I add the gradle wrapper and jars there?

**Comment by ben-manes** [comment]
sure, just keep them all in a single top-level directory, e.g. `examples`, and put each sample underneath it.

File: README.md [hunk 5]
```diff
@@ -85,48 +85,112 @@ The strategy can be specified either on the task or as a system property for ad
...
+       isNonStable(selection.candidate.version)
+   }
+
+   // Example 2: disallow release candidates as upgradable versions from stable versions
+   rejectVersionIf { selection ->
+       isNonStable(selection.candidate.version) && !isNonStable(selection.currentVersion)
+   }
+
+   // Example 3: using the full syntax
+   resolutionStrategy {
```
Comment (ben-manes) [inline README.md]:
```groovy
   resolutionStrategy {
     componentSelection {
       all {
         if (isNonStable(candidate.version) && !isNonStable(currentVersion)) {
           reject('Release candidate')
         }
       }
     }
```

Comment (ghus-raba) [reply README.md]: This is a must as we saw people are quick to copy/paste wrong code from readme. :-)

File: examples/build.gradle [hunk 6]
```diff
@@ -29,20 +29,29 @@ configurations {
...
+
+  rejectVersionIf { selection ->
+    isNonStable(selection.candidate.version) && !isNonStable(selection.currentVersion)
   }
+
+//  resolutionStrategy {
+//    componentSelection { rules ->
+//      rules.all {
+//        if (isNonStable(candidate.version) && !isNonStable(currentVersion)) {
+//          selection.reject('Release candidate')
```
Comment (ghus-raba) [inline examples/build.gradle]: This is same as in readme. If you uncomment this, it would not work, because there is no `selection` variable.

File: src/main/groovy/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentSelectionWithCurrent.groovy [hunk 7]
```diff
@@ -1,13 +1,48 @@
...
 
-  final String currentVersion
+class ComponentSelectionWithCurrent implements ComponentSelection {
+
+  ComponentSelectionWithCurrent(String currentVersion, ComponentSelection delegate) {
+    this.currentVersion = currentVersion
+    this.delegate = delegate
+  }
 
-  @Delegate
```
Comment (ghus-raba) [inline src/main/groovy/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentSelectionWithCurrent.groovy]: Just wondering, what did not work with the delegate annotation? I used it in 0.24.0 in kotlin and could access methods from `ComponentSelection` just fine.

**Comment by jmfayard** [comment]
I pushed 925ad352 to avoid NPE. 

I refactor the smaples:

"examples/groovy" use Gradle 4.10.2 and the buildScripts DSL
"examples/kotlin" uses Gradle 5.6.2 and the plugins DSL

In both cases, `$ ./gradlew install` publishes the plugin to mavenLocal() from which the sample can pick it.

File: README.md (hunk 1 above)
Comment (jmfayard) [reply README.md]: https://github.com/ben-manes/gradle-versions-plugin/pull/340/commits/925ad3524e5d65fc6706f84e5e362aefaa2d4eb2

File: src/main/groovy/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentFilter.groovy [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.github.benmanes.gradle.versions.updates.resolutionstrategy
+
+interface ComponentFilter {
```
Comment (ghus-raba) [inline src/main/groovy/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentFilter.groovy]: Here you could add `@HasImplicitReceiver` like in `org.gradle.api.Action`, and could simplify the usage a little.

File: examples/groovy/gradle/wrapper/gradle-wrapper.properties [hunk 9]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.3-bin.zip
```
Comment (ghus-raba) [inline examples/groovy/gradle/wrapper/gradle-wrapper.properties]: Why did you need to add gradle wrapper to both examples? It could be run from project root like `./gradlew -p examples/groovy dependencyUpdates` and we could keep single wrapper instead of 3.

Comment (jmfayard) [reply examples/groovy/gradle/wrapper/gradle-wrapper.properties]: Oh, very neat, I did that and documented in the README

File: README.md [hunk 10]
```diff
@@ -201,6 +201,19 @@ tasks.withType<DependencyUpdatesTask> {
 
 Note: Do use the `plugins { .. }` syntax if you use the Kotlin DSL.
 
+#### Try out the samples
+
+Have a look at [`examples/groovy`](https://github.com/ben-manes/gradle-versions-plugin/tree/master/examples/groovy) and [`examples/kotlin`](https://github.com/ben-manes/gradle-versions-plugin/tree/master/examples/kotlin)
+
+```bash
+# Pubish the latest version of the plugin to mavenCentral
```
Comment (ghus-raba) [inline README.md]: TYPO: pubLish

Is it supposed to be mavenCentral? Not mavenLocal?

File: src/main/groovy/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentFilter.groovy (hunk 8 above)
Comment (jmfayard) [reply src/main/groovy/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentFilter.groovy]: Oh perfect, that's exactly what I wanted to do but didn't know. 
I think I am now done @ghus-raba

**Comment by ghus-raba** [comment]
Ok, nothing more from me. Just a reminder about this:

> Releasing just requires bumping the version in the `build.gradle` file and creating a Github Release. That makes a new tag, which is then deployed by TravisCI. You're welcome to do this too, if you want 😄.

**Comment by jmfayard** [comment]
Will do it later today. Thanks to both of you. I learned a lot about Java+Kotlin+Groovy interpretability

**Comment by ben-manes** [comment]
and v0.25 is live =)
